### ISSUE TITLE
Typing non-initial hyphen in name shows error message too soon

### ISSUE THREAD

**Issue by nottrobin** [canonical-web-and-design/build.snapcraft.io#347]
1\. Add a repo.
2\. Choose “Not registered”.
3\. Type “-”.
4\. Delete the “-” and type “flim-”.
5\. Continue typing, “flam-”, and choose “Register name”.

What happens:

3\. An error message appears, “Sorry the name can't start or end with a hyphen.”
4\. The same error message appears.
5\. The same error message appears.

What should happen:

3\. The error message appears.
4\. The error message _does not_ appear, because you may not have finished typing yet.
5\. The error message appears.

[From the specification](https://docs.google.com/document/d/1koTwI3cRr2S2vj5zHKKbR7aWxLRfShDVq2gHQ29cDyU/edit#bookmark=id.9r0spyi9a925):
> Register name” should be disabled whenever:
> - the name field is empty, or
> - “I accept the Developer Programme Agreement” is present but unchecked, or
> - the first character is a hyphen (with the error message “Sorry, the name can’t start or end with a hyphen.”).
> 
> (#347: The button should not be disabled because the *last* character is a hyphen, because unfocusing the field — most obviously, by clicking the button — is the moment we explain that that isn’t allowed. Explaining it initially would be too wordy, and explaining it on input would display an error message even for legitimate hyphens.)
>…
> If registration fails, the field and button should become enabled again, and the caption should change to one of:
> - Sorry, the name can’t start or end with a hyphen.
>
>…

---

Copied from Trello card: https://trello.com/c/BOG6kcBh/259-typing-non-initial-hyphen-in-name-shows-error-message-too-soon

**Comment by nottrobin** [comment]
@matthewpaulthomas I don't think this is quite up-to-date with what we discussed in the workshop, but I'm not sure I know exactly what we decided. Are we supposed to never disable the button? Also, should we only be doing any validation on blue? Could you update it with your understanding, or link to the relevant part of the copy doc?

**Comment by matthewpaulthomas** [comment]
@nottrobin Updated. The relevant part of the spec could be found by searching for “hyphen”, or even “#347” (though the latter works for few issues so far). Let me know if you have any suggestions for making the spec easier to navigate.

I’m sorry I don’t know what “validation on blue” means.

**Comment by nottrobin** [comment]
@matthewpaulthomas sorry, "validation on blur" - as in when the field loses focus.

**Comment by matthewpaulthomas** [comment]
Ah, of course. Yes, the validation for the “can’t end with a hyphen” rule, specifically, should take place on blur (and also on submit). Hopefully this is clearer than my attempt to explain it in the workshop:

We can’t ignore a non-initial hyphen altogether, because we don’t know whether other characters will follow.

We can’t show an error message as soon as you enter the hyphen, for the same reason.

Showing an error message on a timeout would have the same problem, just unpredictably. If you’re choosing a snap name, you might well pause in the middle to think, and thinking shouldn’t be an error.

The next best thing is to show the error message on blur.

However, in this form (assuming that you’ve accepted the Developer Agreement in the past), there’s no obvious reason for you _to_ blur the field … except by clicking the button.

Therefore, for this particular rule, we should keep the button enabled even though we know the registration isn’t going to work. The button is bait for the blur.

Other rules for this field — the name can’t be empty, and the name can’t start with a hyphen — aren’t subject to this logic, so the button should be disabled immediately whenever those requirements aren’t met.

**Comment by Greggless** [comment]
### related: #963

**Comment by bartaz** [comment]
@Greggless I don't think they are related. This one is about registering a name in the store, the #963 is about repository having dash in the beginning of the name.

**Comment by solazio** [comment]
Not relevant for snapcraft.io, closing
